pkg/volume/scaleio/sio_client.go

```diff
@@ -0,0 +1,430 @@
+	diskIDPath := "/dev/disk/by-id"
+	files, _ := ioutil.ReadDir(diskIDPath)
+	r, _ := regexp.Compile(`^emc-vol-\w*-\w*$`)
+
+	for _, f := range files {
+		matched := r.MatchString(f.Name())
+		if matched {
+			// remove emec-vol- prefix to be left with concated mdmID-volID
+			mdmVolumeID := strings.Replace(f.Name(), "emc-vol-", "", 1)
+			devPath, err := filepath.EvalSymlinks(fmt.Sprintf("%s/%s", diskIDPath, f.Name()))
+			if err != nil {
+				return nil, err
+			}
+			if volumeID, ok := mdmMap[mdmVolumeID]; ok {
+				volumeMap[volumeID] = devPath
+			}
+		}
+	}
+	return volumeMap, nil
+}
+
+// WaitForAttachedDevice sets up a timer to wait for an attached device to appear in the instance's list.
+func (c *sioClient) WaitForAttachedDevice(token string) (string, error) {
+	if token == "" {
+		return "", fmt.Errorf("invalid attach token")
+	}
+
+	// wait for attach.Token to show up in local device list
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	timer := time.NewTimer(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			devMap, err := c.Devs()
+			if err != nil {
+				return "", err
+			}
+			for _, v := range devMap {
+				if v == token {
+					return devMap[token], nil
+				}
+			}
+		case <-timer.C:
+			return "", fmt.Errorf("volume attach timeout")
+		}
+	}
+}
+
+// waitForDetachedDevice waits for device to be detached
+func (c *sioClient) WaitForDetachedDevice(token string) error {
+	if token == "" {
+		return fmt.Errorf("invalid detach token")
+	}
+
+	// wait for attach.Token to show up in local device list
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	timer := time.NewTimer(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			devMap, err := c.Devs()
+			if err != nil {
+				return err
+			}
+			// cant find vol id, then ok.
+			if _, ok := devMap[token]; !ok {
+				return nil
+			}
+		case <-timer.C:
+			return fmt.Errorf("volume detach timeout")
+		}
+	}
+}
+
+// ***********************************************************************
+// Little Helpers!
+// ***********************************************************************
+func (c *sioClient) findSystem(sysname string) (sys *siotypes.System, err error) {
+	if c.sysClient, err = c.client.FindSystem("", sysname, ""); err != nil {
+		return nil, err
+	}
+	systems, err := c.client.GetInstance("")
+	if err != nil {
+		return nil, err
+	}
+	for _, sys = range systems {
+		if sys.Name == sysname {
+			return sys, nil
+		}
+	}
+	return nil, errors.New("instance not found")
+}
+
+func (c *sioClient) findProtectionDomain(pdname string) (*siotypes.ProtectionDomain, error) {
+	c.pdClient = sio.NewProtectionDomain(c.client)
+	if c.sysClient != nil {
+		protectionDomain, err := c.sysClient.FindProtectionDomain("", pdname, "")
+		if err != nil {
+			return nil, err
+		}
+		c.pdClient.ProtectionDomain = protectionDomain
+		return protectionDomain, nil
+	}
+	return nil, errors.New("system not set")
```

Comment:
the errors are messed up, should be ProtectionDomain rather than system. So are those below.